src/transforms/kubernetes/mod.rs

```diff
@@ -1 +1,568 @@
+        // Main idea is to have a background task which will premptively
+        // acquire metadata for all pods on this node, and with it maintaine
+        // a map of extracted metadata.
+
+        // Detemine Node's name of whose Pod's metadata we will watch for.
+        let node = if let Some(node) = self.node_name.clone() {
+            node
+        } else {
+            std::env::var(NODE_NAME_ENV)
+                .map_err(|_| BuildError::MissingNodeName { env: NODE_NAME_ENV })?
+        };
+
+        // Construct WatchClient
+        let wc_config = ClientConfig::in_cluster(node, cx.resolver()).context(WatchClientBuild)?;
+        let watch_client = wc_config.build().context(WatchClientBuild)?;
+
+        // Construct MetadataClient
+        let (reader, writer) = evmap::new();
+        let metadata_client = MetadataClient::new(
+            self.fields
+                .clone()
+                .unwrap_or(DEFAULT_FIELDS.iter().map(Clone::clone).collect()),
+            self.cache_ttl
+                .map(Duration::from_secs)
+                .unwrap_or(DEFAULT_CACHE_TTL),
+            self.max_retry_timeout
+                .map(Duration::from_secs)
+                .unwrap_or(DEFAULT_MAX_RETRY_TIMEOUT),
+            writer,
+            watch_client,
+        )?;
+
+        // Run background task
+        cx.executor().spawn_std(async move {
+            match metadata_client.run().await {
+                Ok(()) => (),
+                Err(error) => error!(message = "Stopped updating Pod metadata.", reason = %error),
+            }
+        });
+
+        // Construct transform
+        Ok(Box::new(KubernetesPodMetadata {
+            metadata: reader,
+            pod_uid: self
+                .pod_uid
+                .clone()
+                .map(Into::into)
+                .unwrap_or(POD_UID.clone()),
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "kubernetes_pod_metadata"
+    }
+}
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("Failed building WatchClient: {}", source))]
+    WatchClientBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display("Failed building watch stream: {}", source))]
+    WatchStreamBuild {
+        source: self::watch_client::BuildError,
+    },
+    #[snafu(display(
+        "Missing environment variable {:?} containing node name `spec.nodeName`.",
+        env
+    ))]
+    MissingNodeName { env: &'static str },
+    #[snafu(display("max_retry_timeout must be larger than 0."))]
+    TooSmallMaxRetryTimeout,
+}
+
+/// Vlient which watches for Pod metadata changes, extracts fields,
+/// and writes them to the metadata map.
+struct MetadataClient {
+    fields: Vec<Field>,
+    metadata: WriteHandle<Bytes, Box<(Atom, FieldValue)>>,
```

Comment:
It is, `WriteHandle` requires that its value implement `ShallowCopy` which is unsafe to implement so we are using `Box` as a wrapper since `ShallowCopy` is implemented for it.

Reply:
Ah ok I forgot there is a bytes feature and it does this for bytes :D